File: java-midterm/src/main/java/com/pht/service/CategoryService.java
package com.pht.service;

import java.util.List;

import com.pht.dto.request.CategoryRequest;
import com.pht.entity.CategoryEntity;

public interface CategoryService {
	List<CategoryEntity> getAll();

	CategoryEntity create(CategoryRequest categoryRequest);

	CategoryEntity update(CategoryRequest categoryRequest, long id);

	void delete(long id);
}

File: midterm/src/app/services/fetch/category.service.ts
import { Injectable } from '@angular/core';
import { BaseService } from './base.service';
import { Observable } from 'rxjs';
import { CategoryResponse } from '../../models/category';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  constructor(private readonly baseService: BaseService) { }

  getAll(): Observable<CategoryResponse[]> {
    return this.baseService.get<CategoryResponse[]>('/api/categories/');
  }

  create(formData: FormData): Observable<CategoryResponse> {
    return this.baseService.post<CategoryResponse>('/api/categories/', formData, true);
  }

  update(id: number, formData: FormData): Observable<CategoryResponse> {
    return this.baseService.put<CategoryResponse>(`/api/categories/${id}/`, formData, true);
  }

  delete(id: number): Observable<any> {
    return this.baseService.delete<any>(`/api/categories/${id}/`);
  }
}
